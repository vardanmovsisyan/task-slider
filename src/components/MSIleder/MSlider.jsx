import React, { useState } from "react";

import classes from "./MSlider.module.scss";

function MSlider(props) {
  const [activeElement, SetActiveElement] = useState(0);
  const [previous, SetPreviousElement] = useState(0);
  const [duration, SetDurationElement] = useState(0);
  const [lastClick, SetLastClickTime] = useState(0);
  const [elements] = useState(props.elements);
  const [animationDuration] = useState(props.duration);

  function multipleClicksCheckPrevent(){
      if (+new Date() - 1000 * 0.65 * animationDuration < lastClick){
          return false;
      }
      return true;
  }
  function handleSlideChange(activeElement,_activeElement) {
      SetPreviousElement(activeElement);
      SetActiveElement(_activeElement);
      SetLastClickTime(+new Date());
      if (typeof props.callback === "function"){
          props.callback(elements.length, _activeElement + 1);
      }
  }
  function slideLeft() {
    if(!multipleClicksCheckPrevent()){
        return false;
    }
    let _activeElement = activeElement;
    if (--_activeElement === -1) {
      _activeElement = elements.length - 1;
    }
    SetDurationElement(0);
    handleSlideChange(activeElement,_activeElement);
  }
  function slideRight() {
    if(!multipleClicksCheckPrevent()){
        return false;
    }
    let _activeElement = activeElement;
    if (++_activeElement === elements.length) {
      _activeElement = 0;
    }
    SetDurationElement(1);
    handleSlideChange(activeElement,_activeElement);
  }

  let activeClassName=duration===0?classes.activeFromLeft:classes.activeFromRight;
  let previousClassName=duration===0?classes.previousLeft:classes.previousRight;

  return (
    <ul className={classes.MSlider}>
      <li onClick={slideLeft} className={classes.slideLeft} data-testid="slideLeftButton">
        {" "}
        {"<"}{" "}
      </li>
      {elements.map((elem, index) => {
        return (
          <li
            style={{ animationDuration: animationDuration + "s" }}
            className={
              classes.elements +
              " " +
              (index === activeElement
                ? activeClassName
                : index === previous
                  ? previousClassName
                  : classes.passive)
            }
            key={index}
            data-testid={(index===activeElement)?'activeElement':'passiveElement-'+index}
          >
            <div className={classes.MImageContainer}>
              <img src={elem} alt="" />
            </div>
          </li>
        );
      })}

      <li onClick={slideRight} className={classes.slideRight} data-testid="slideRightButton">
        {" "}
        {">"}{" "}
      </li>
    </ul>
  );
}

export default MSlider;
