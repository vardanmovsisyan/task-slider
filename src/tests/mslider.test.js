'use strict';

import React from 'react';
import MSlider from "../components/MSIleder/MSlider";
import renderer from 'react-test-renderer';
import {render, cleanup} from '@testing-library/react';

const elements = [
    "https://www.armeniatravel.am/wcit/V.2.0/images/Amberd/Amberd_1.jpg",
    "https://www.armeniatravel.am/wcit/V.2.0/images/Sevan_Dilijan/Sevan.jpg",
    "https://www.armeniatravel.am/wcit/V.2.0/images/Amberd/Amberd_day_4.jpg",
    "https://www.armeniatravel.am/wcit/V.2.0/images/Amberd/Amberd_2.jpg"
];

describe('mslider', () => {
    afterEach(cleanup);

    it('renders correctly', () => {
        const tree = renderer
            .create(<MSlider elements={elements}
                             duration={2}
                             callback={function(a, b) {
                                 console.log(a, b);
                             }}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('arrow right handler event set activeElement from 0 to 1 when invoked', () => {
        const {
            getByTestId
        } = render(<MSlider elements={elements}
                            duration={2}
                            callback={function(a, b) {
                                console.log(a, b);
                            }}/>);
        const rightArrowButton = getByTestId('slideRightButton');
        rightArrowButton.click();
        expect(getByTestId('activeElement').querySelector('img').getAttribute('src')).toBe(elements[1]);
    });

    it('arrow left handler event set activeElement from 0 to the number matching elements length minus 1 when invoked', () => {
        const {
            getByTestId
        } = render(<MSlider elements={elements}
                            duration={2}
                            callback={function(a, b) {
                                console.log(a, b);
                            }}/>);
        const leftArrowButton = getByTestId('slideLeftButton');
        leftArrowButton.click();
        expect(getByTestId('activeElement').querySelector('img').getAttribute('src')).toBe(elements[elements.length-1]);
    });
});
