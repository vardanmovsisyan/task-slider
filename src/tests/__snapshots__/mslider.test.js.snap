// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mslider renders correctly 1`] = `
<ul>
  <li
    data-testid="slideLeftButton"
    onClick={[Function]}
  >
     
    &lt;
     
  </li>
  <li
    className="undefined undefined"
    data-testid="activeElement"
    style={
      Object {
        "animationDuration": "2s",
      }
    }
  >
    <div>
      <img
        alt=""
        src="https://www.armeniatravel.am/wcit/V.2.0/images/Amberd/Amberd_1.jpg"
      />
    </div>
  </li>
  <li
    className="undefined undefined"
    data-testid="passiveElement-1"
    style={
      Object {
        "animationDuration": "2s",
      }
    }
  >
    <div>
      <img
        alt=""
        src="https://www.armeniatravel.am/wcit/V.2.0/images/Sevan_Dilijan/Sevan.jpg"
      />
    </div>
  </li>
  <li
    className="undefined undefined"
    data-testid="passiveElement-2"
    style={
      Object {
        "animationDuration": "2s",
      }
    }
  >
    <div>
      <img
        alt=""
        src="https://www.armeniatravel.am/wcit/V.2.0/images/Amberd/Amberd_day_4.jpg"
      />
    </div>
  </li>
  <li
    className="undefined undefined"
    data-testid="passiveElement-3"
    style={
      Object {
        "animationDuration": "2s",
      }
    }
  >
    <div>
      <img
        alt=""
        src="https://www.armeniatravel.am/wcit/V.2.0/images/Amberd/Amberd_2.jpg"
      />
    </div>
  </li>
  <li
    data-testid="slideRightButton"
    onClick={[Function]}
  >
     
    &gt;
     
  </li>
</ul>
`;
